Add static grid gutters #8508

ncoden opened the pull request:
**Add static (unresponsive) grid gutters**
See: https://github.com/zurb/foundation-sites/issues/8508
Use: `.gutter-{size}` class on a row

Example:

``` html
<div class="row gutter-small"> <!-- Gutters always have a "small" width -->
  <div class="large-6 columns">...</div>
  <div class="large-6 columns">...</div>
  ...
</div>
```

**Standardize grid gutters functions, mixins and arguments.**

Adds/Removes:
- Add `@function grid-column-gutter( {breakpoint} )`: Get a gutter size for a given breakpoint.
- Add `@mixin grid-column-size( {px or breakpoint} )`: Set the gutters on a column.
- Depreciate `@mixin grid-column-uncollapse`. Use `grid-column-gutter` instead.
- Use these new features in all `(flex-)grid` mixin & functions. Use `@function grid-column-gutter` instead of modifying `$grid-column-gutter` to support its possible single value.

Other changes:
- Argument naming: use `$gutter` for a single size, and `$gutters` for a size map.
- Fix `$gutter` (now `$gutters`) argument descriptions.

_Required for the previous feature.
Anyway, all the "groups of properties" of a component (i.e. `size`, `align`, `color`, or specific behavior like `collapse`, `expand`...) should have their own `@mixin`, to allow a component customization in responsive/semantic cases_

ncoden committed:
Add static grid gutters #8508

**Add static (unresponsive) grid gutters**
See: https://github.com/zurb/foundation-sites/issues/8508
Use: `. row.gutter-{size}`

Example:

```
<div class="row gutter-small"> <!-- Gutters has always "small" width -->
  <div class="large-6 columns">...</div>
  <div class="large-6 columns">...</div>
  ...
</div>
```

**Add grid-column-gutter mixin**
Set the gutters on a column.
Use: `@include grid-column-gutter($width);`

kball commented:
@ncoden I like the direction this is going. You should be aware though that $grid-column-gutter currently supports setting it to a single (scalar) value as well as the responsive approach... I think this would break in that case.  Probably in that situation these should just degenerate to the same gutter, yes?

andycochran commented:
@kball Ah, right! This'll need to account for if you're not using responsive gutters (e.g. `$grid-column-gutter` is a single value).

ncoden commented:
:+1: hum. Actually, if `$grid-column-gutter` is a single value, it is transformed to a `(small: $value)` map.
So it would generate a useless `.gutter-small` class.

ncoden commented:
It's dirty because we're assuming there is always a smallest breakpoint called "small".

So I purpose to not modify `$grid-column-gutter`, and create:
```scss
@function grid-column-gutter( $breakpoint )
```
that return *the gutter* if it is a single value, or *the gutter for a given breakpoint* if it is a map.

andycochran commented:
You can use `nth(map-keys($breakpoints),1)` to get the smallest (zero) "breakpoint."

ncoden commented:
> You can use nth(map-keys($breakpoints),1) to get the smallest (zero) "breakpoint."
Can there be no breakpoint ? Does the "zero breakpoint" always exists and be zero ?

I also see a problem with `grid-column-uncollapse`:
```scss
@mixin grid-column-uncollapse($gutter: $grid-column-gutter) { ... }
```
- `$grid-column-gutter` is always a map, and is used as padding. We must remove this bugged default value`
- So `$grid-column-uncollapse` is an alias of `$grid-column-gutter`
- And `$grid-column-collapse` is an alias of `$grid-column-gutter(0)`
- So it would be better to only have:

```scss
@include grid-column-gutter( {size_in_px} );
@include grid-column-gutter( {breakpoint} ); // "small", "medium"...
```

And maybe
```scss
@include grid-column-gutter( collapse ); // alias for 0
```

*CSS classes keep unchanged*

andycochran commented:
@ncoden Yeah. A zero breakpoint is required as the first key/value of the `$breakpoints` map. 

But [the error](https://github.com/zurb/foundation-sites/blob/12317b1854283b3a2708d586f1141d523684f8cd/scss/util/_breakpoint.scss#L24) is misleading. 

This:
```scss
@error 'Your smallest breakpoint (defined in $breakpoints) must be set to "0".';
```

Should probably be:
```scss
@error 'The first key in the $breakpoints map must have a value of "0".';
```

And a note should also be added to [the docs page](http://foundation.zurb.com/sites/docs/media-queries.html#changing-the-breakpoints).

DaSchTour commented:
BTW, since #8318 is merged you can use $-zf-zero-breakpoint to get the smallest breakpoint.

ncoden commented:
@andycochran @DaSchTour @kball Ok, I will add theses changes.

ncoden committed:
Standardize grid gutters

Standardize grid gutters functions, mixins and arguments.

Adds/Removes:
- Add `grid-column-gutter` function: Get a gutter size for a given
breakpoint.
- Allow `grid-column-size` mixin to take breakpoint name as argument
- Remove `grid-column-uncollapse` mixin. Use `grid-column-size` instead.
- Use these new features in all `(flex-)grid` mixin & functions. Use
`grid-column-gutter` function instead of modify `$grid-column-gutter`
variable

Minor changes:
- Use `$gutter` for a single size, and `$gutters` for a size map

ncoden commented:
I done these changes, and standardize/factorized all the gutter functions/mixins/uses. See all the changes in this PR description.

I have two questions:
- Should I update the documentation ? (docs/pages/grid.md)
- Should I keep `@mixin grid-column-uncollapse` deleted, or would it be better to depreciate it (redirection to `@mixin grid-column-size` and throw a warning) ?

kball commented:
@ncoden I prefer depreciations than outright deletions, so that sounds like a good step.  Updating the documentation would be great.  @andycochran can you try this out and see if the code changes are good to merge?

kball commented:
@ncoden looks like this branch has gotten stale and needs some conflicts resolved before it can be merged.  @andycochran can you give this trial run?

ncoden commented:
Ah sorry, i forgot this PR ;)
I rebase it...

ncoden committed:
Fix grid gutters

Fixs:
- Depreciate `grid-column-uncollapse` and `grid-col-uncollapse` instead
of delete it
- Fix `.row.{breakpoint}-uncollapse` (wrong mixin call)

ncoden commented:
@kball @andycochran  It's rebased (and fixed). :)

ncoden committed:
Fix grid gutters (2)

Fixs:
- Fix `grid-column-gutter` misnamed passed parameter
- Add missing `grid-col-gutter` default parameter

kball commented on scss/grid/_column.scss:
```diff
@@ -87,12 +75,12 @@
 /// Creates a grid column row. This is the equivalent of adding `.row` and `.column` to the same element.
 ///
-/// @param {Number} $gutter [$grid-column-gutter] - Width of the gutters on either side of the column row.
+/// @param {Mixed} $gutter [$grid-column-gutter] - Width of the gutters on either side of the column row. Refer to the `grid-column-gutter()` function to see possible values.
```
comment @param name should prob get updated to $gutters

ncoden replied:
Thanks

ncoden replied:
@kball Fixed

kball commented:
@andycochran @brettsmason @JeremyEnglert as our scss expert yetinauts, does this seem ready to you?

ncoden committed:
Fix grid-column-row parameter in doc
